Remove some `GCHandle`s from `SendHeaders()`

AaronRobinsonMSFT opened the pull request:
## Description

Fixes https://github.com/dotnet/aspnetcore/issues/40215

This introduces a new type, `UnmanagedBufferAllocator`, that manages unmanaged block allocations. The intent here is to remove the allocation of handles and temporary managed arrays and replace them with native memory.

Instead of using an unmanaged memory allocator like `NativeMemory`, the `GC.AllocateUninitializedArray()` could be used to create pinned memory. I've rejected this because it puts the onus on the GC to collect the memory whereas with the unmanaged memory it can be freed precisely when it is no longer needed. Since the pinned heap is presently considered only on a gen2 collection it leaves this unused memory around far longer than needed and creates unnecessary pressure.

AaronRobinsonMSFT committed:
Remove some GCHandles from SendHeaders()

This introduces a new IBufferWriter implementation
that manages unmanaged block allocations. The
intent here was to remove the allocation of handles
amd temporary managed arrays allocations and
and replace them with native memory.

AaronRobinsonMSFT committed:
Comments and style

AaronRobinsonMSFT commented:
/benchmark list

AaronRobinsonMSFT committed:
Change concrete type to interface in signature

davidfowl commented:
Http.sys isn't in this list. I recently did a perf change to http.sys and had to build locally and run a benchmark. We should add the JSON one.

cc @sebastienros @adityamandaleeka @Tratcher

AaronRobinsonMSFT committed:
Match the semantics of the GetBytes() API

davidfowl commented on src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:
```diff
@@ -27,4 +28,21 @@ internal static byte[] GetBytes(string myString)
+        writer.Advance(written);
+
+        // The resulting Span should not include the null terminator in its length.
+        return buffer.Slice(0, written - 1);
```
C# in 2022 bruh 😄 

```suggestion
        return buffer.Slice[..^1];
```

AaronRobinsonMSFT committed:
Fix use of wrong class.

davidfowl commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -289,45 +302,57 @@ private void EnsureResponseStream()
         uint bytesSent;
-        List<GCHandle>? pinnedHeaders = SerializeHeaders(isOpaqueUpgrade);
-        try
+        using UnmanagedBufferWriter writer = new(4096);
```
I'm not sure this can be disposed until the async operation is over.

AaronRobinsonMSFT replied:
I thought so too, but I don't see how this worked with the previous logic then because all the GC handles were disposed of in the finally block below. I need to reexamine the logic to see what I am missing.

AaronRobinsonMSFT replied:
Reverting my changed does make this pass, but it very odd how this works because of the aforementioned `finally` block that will always free the pinned handled and thus permit the pointers to move.

https://github.com/dotnet/aspnetcore/blob/1b53b22a866eb4bc58f2d5552db7f3ac08f1a952/src/Servers/HttpSys/src/RequestProcessing/Response.cs#L379-L382

Another oddity is the arguments passed to `HttpSendHttpResponse`. The [official documentation](https://learn.microsoft.com/windows/win32/api/http/nf-http-httpsendhttpresponse) states the following for the `BytesSent` argument:

> Optional. A pointer to a variable that receives the number, in bytes, sent if the function operates synchronously.
>
> When making an asynchronous call using pOverlapped, set pBytesSent to NULL. Otherwise, when pOverlapped is set to NULL, pBytesSent must contain a valid memory address and not be set to NULL.

However, it is always non-NULL in the current code path, so perhaps this is a no-op? Unclear at present.

I can make this work but I'd like to hear from someone who knows this code path well. The current GCHandle approach seems to work but isn't obvious why. @Tratcher do you have any thoughts on the above?

AaronRobinsonMSFT replied:
I started over. This was my mistake in how to deref an empty `Span<T>` and there was code that was assuming zero initialization which wasn't done automatically with the native allocator.

AaronRobinsonMSFT committed:
Use MemoryMarshal.GetReference to deref
an empty Span<>.
Zero initialize count member.

jkotas commented on src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:
```diff
@@ -27,4 +28,21 @@ internal static byte[] GetBytes(string myString)
+    {
+        // Compute the maximum amount of bytes needed for the given string.
+        // Include an extra byte for the null terminator.
+        Span<byte> buffer = writer.GetSpan(Encoding.GetMaxByteCount(myString.Length) + 1);
```
This (and everything else in this file) should use Encoding.UTF8 directly, and the static encoding cache should be deleted.

Using Encoding.UTF8 directly allows better devirtualization.

AaronRobinsonMSFT replied:
The static allocation above sets `encoderShouldEmitUTF8Identifier: false`, whereas `Encoding.UTF8` sets it to `encoderShouldEmitUTF8Identifier: true`. I can however change the type to be `UTF8Encoding` instead.

jkotas replied:
`encoderShouldEmitUTF8Identifier` matters only when the encoding is used for streaming (ie you get encoder and feed a stream of data into the encoder). None of the encoding uses in this type are streaming. `Encoding.UTF8` will work just fine here. 

> I can however change the type to be UTF8Encoding instead.

This helps a little, but it still worse than using `Encoding.UTF8` directly.

jkotas commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs:
```diff
@@ -0,0 +1,95 @@
+    /// <returns>The allocated memory.</returns>
+    public byte* Alloc(int size)
+    {
+        var alloc = (byte*)NativeMemory.AlignedAlloc((nuint)size, 8);
```
I do not think you need to specify explicit alignment here. The default alignment of the NativeMemory allocator should be enough.

jkotas commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs:
```diff
@@ -0,0 +1,95 @@
+    public byte* Alloc(int size)
+    {
+        var alloc = (byte*)NativeMemory.AlignedAlloc((nuint)size, 8);
+        _allocations.Add((IntPtr)alloc);
```
`_allocation.Add` throws, the unmanaged memory chunk is going to leak.

It may be better to use first pointer in the native memory block for a linked list, so that you do not need maintain `List<T>` with the pointers. It would avoid this problem as side-effect.

jkotas commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs:
```diff
@@ -0,0 +1,95 @@
+        }
+        else
+        {
+            NewBlock();
```
This does not look right. Advance always only commit what was written. It should not need to ever allocate a new chunk.

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -527,111 +527,89 @@ private static bool Matches(string knownValue, StringValues input)
+                    knownHeaderInfo[_nativeResponse.ResponseInfoCount].Type = HttpApiTypes.HTTP_RESPONSE_INFO_TYPE.HttpResponseInfoTypeMultipleKnownHeaders;
+                    knownHeaderInfo[_nativeResponse.ResponseInfoCount].Length = (uint)sizeof(HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS);
+
+                    var header = (HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS*)writer.Alloc(sizeof(HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS));
```
It would be better if this tried to used memory from the pooled block instead of a separate allocation.

jkotas commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs:
```diff
@@ -0,0 +1,95 @@
+/// An <see cref="IBufferWriter{T}"/> implementation for writing to
+/// unmanaged blocks of memory.
+/// </summary>
+internal sealed unsafe class UnmanagedBufferWriter : IBufferWriter<byte>, IDisposable
```
Does this really need to go through `IBufferWriter<byte>` abstraction? I think this can be (ref) struct and it does not need to implement `IBufferWriter<byte>` interface. It would save allocation and some interface virtual calls.

davidfowl replied:
Agree we should be able to get away with something like https://github.com/dotnet/aspnetcore/blob/75ada08ac7763cf174c05fbe0755e184c0ad9737/src/Shared/ServerInfrastructure/BufferWriter.cs. You can avoid the IBufferWriter calls for encoding as well and just pass the `Span<byte>`

jkotas commented on src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:
```diff
@@ -27,4 +28,21 @@ internal static byte[] GetBytes(string myString)
     {
```
Is this `GetBytes` method still used anywhere after these changes?

AaronRobinsonMSFT replied:
Yes it is used in a few places. I am working on removing it but wanted a review of the current approach first.

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -289,45 +302,57 @@ private void EnsureResponseStream()
-            byte[] reasonPhraseBytes = HeaderEncoding.GetBytes(reasonPhrase);
-            fixed (byte* pReasonPhrase = reasonPhraseBytes)
+        Span<byte> reasonPhraseBytes = HeaderEncoding.GetBytes(reasonPhrase, writer);
+        fixed (byte* pReasonPhrase = reasonPhraseBytes)
```
This does not need to use `fixed`. It can just be `AsPointer` like in the other places.

It may be better if GetBytes signature is `byte* GetBytes(string s, out int length, UnmanagedBufferWriter writer)`  because all callers want the pointer at the end. It would avoid duplication of the `Unsafe.AsPointer(ref Memory.Marshal(...` dance everywhere.

jkotas replied:
GetBytes can be even instance method on UnmanagedBufferWriter.

davidfowl commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -527,111 +527,89 @@ private static bool Matches(string knownValue, StringValues input)
+                    headerValue = headerValues[0] ?? string.Empty;
+                    bytes = HeaderEncoding.GetBytes(headerValue, writer);
+                    pKnownHeaders[lookup].RawValueLength = (ushort)bytes.Length;
+                    pKnownHeaders[lookup].pRawValue = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(bytes));
```
Is this the cleanest way to do this?

AaronRobinsonMSFT replied:
As in get to the pointer within a `Span<T>`? Yes. This is a very advanced scenario, mostly used in interop, so making it cleaner isn't worth a lot of effort.

davidfowl commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs:
```diff
@@ -0,0 +1,95 @@
+    /// </summary>
+    /// <param name="size">The amount of memory to allocate in bytes.</param>
+    /// <returns>The allocated memory.</returns>
+    public byte* Alloc(int size)
```
This should be private right?

AaronRobinsonMSFT replied:
Nope. At present it is used directly. I am reworking this based on https://github.com/dotnet/aspnetcore/pull/44561#discussion_r996373072.

davidfowl commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs:
```diff
@@ -0,0 +1,95 @@
+internal sealed unsafe class UnmanagedBufferWriter : IBufferWriter<byte>, IDisposable
+{
+    private readonly int _blockSize;
+    private readonly List<IntPtr> _allocations = new();
```
If this was a ref struct, then this could also be stack allocated for 99% of cases.

AaronRobinsonMSFT replied:
That is an interesting idea. My concern there is taking up so much stack space. How much would be needed in practice? 256? 512? I wouldn't want to consume much more than 512.

AaronRobinsonMSFT replied:
Oh, do you mean the storing of the allocation chain? That is being addressed with https://github.com/dotnet/aspnetcore/pull/44561#discussion_r996372383, which is how I would do it in C/C++, but I wasn't entirely sure how appealing that would be in C#.

davidfowl replied:
> Oh, do you mean the storing of the allocation chain?

Yep, I was thinking about the list allocation to store the number of pointers. Since this is all synchronous code maybe the stack works well for that.

> That is being addressed with https://github.com/dotnet/aspnetcore/pull/44561#discussion_r996372383, which is how I would do it in C/C++, but I wasn't entirely sure how appealing that would be in C#.

Does it mean use a part of the native memory itself as a linked list node? That sounds reasonable.

AaronRobinsonMSFT replied:
> Does it mean use a part of the native memory itself as a linked list node? That sounds reasonable.

Yep.

jkotas replied:
If the total memory needed by these headers is reasonably small in typical cases, it may be worth it to allocate the first block on stack and continue allocating on native only when the stack allocated block is not sufficient.

@davidfowl Do you have an estimate how much memory is needed for these headers in 80% case?

davidfowl replied:
headers can get big (auth tokens), but in benchmarks we can make it more pay as you go (or non-browser requests). I'm thinking 256-512b of scratch space and then switch to the native heap from there.

AaronRobinsonMSFT replied:
Due to how this is now being used, I don't see how 256-512 is going to move the needle. From my basic investigation this seems to be tripped very often. I think we should go with what we have right now and reconsider in the future. I really don't want to play the stackalloc game on these paths right now given all the other changes in this PR.

AaronRobinsonMSFT committed:
Review comments.

Remove use of IBufferWriter<T> and replace with
struct with methods for scenarios.
Move encoding logic to use Encoding.UTF8 directly.

AaronRobinsonMSFT committed:
Respond to review feedback.
Plumb UnmanagedBufferAllocator through APIs.
Remove many GCHandle allocations.
Add UnmanagedBufferAllocator field to ResponseStreamAsyncResult.

AaronRobinsonMSFT committed:
Remove unused using.

AaronRobinsonMSFT committed:
Remove unused usings.

jkotas commented on src/Servers/HttpSys/src/HttpSysListener.cs:
```diff
@@ -310,103 +309,78 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+
+        bytes = allocator.GetHeaderEncodedBytes(statusDescription, out bytesLength);
+        httpResponse.Response_V1.pReason = bytes;
+        httpResponse.Response_V1.ReasonLength = (ushort)bytesLength;
```
Given that the length is passed around everywhere, does httpsys require the header-encoded bytes need to be zero-terminated? We can save a bit by skipping the zero-termination.

AaronRobinsonMSFT replied:
They do not. I am simply terminating because that helps with viewing the buffer. I prefer terminating as a matter of consistency in the buffer logic.

jkotas commented on src/Servers/HttpSys/src/HttpSysListener.cs:
```diff
@@ -310,103 +309,78 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+
+        int pContentLengthLength = 1;
+        byte* pContentLength = stackalloc byte[pContentLengthLength];
+        pContentLength[0] = (byte)'0';
```
This string is not zero-terminated.

AaronRobinsonMSFT replied:
It doesn't need to be. I can add it to be consistent though.

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -277,10 +278,25 @@ private void EnsureResponseStream()
         HttpApiTypes.HTTP_FLAGS flags,
         bool isOpaqueUpgrade)
+    {
+        fixed (HttpApiTypes.HTTP_DATA_CHUNK* chunks = dataChunks)
```
Merge the internal and private SendHeaders together? Or is there a reason for them to be split (except the cosmetic reason of an extra indentation)?

AaronRobinsonMSFT replied:
This was based on an earlier thought I had. It doesn't provide any value anymore.

jkotas commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferAllocator.cs:
```diff
@@ -0,0 +1,156 @@
+    /// <returns>A pointer to the reserved memory.</returns>
+    public T* Alloc<T>(int count) where T : unmanaged
+    {
+        int toAlloc = count *sizeof(T);
```
```suggestion
        int toAlloc = checked(count * sizeof(T));
```
to avoid silent integer overflow that will lead to a buffer overrun.

jkotas commented on src/Servers/HttpSys/src/HttpSysListener.cs:
```diff
@@ -310,103 +309,78 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+                // Add Value
+                string headerValue = authChallenges[headerValueIndex];
+                bytes = allocator.GetHeaderEncodedBytes(headerValue, out bytesLength);
+                nativeHeaderValues[header->KnownHeaderCount].RawValueLength = (ushort)bytesLength;
```
These casts to `ushort` should be checked to avoid silent integer overflows (unless there is a strong guarantee somewhere that these header values are always going to fit).

jkotas commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferAllocator.cs:
```diff
@@ -0,0 +1,156 @@
+
+        // Compute the maximum amount of bytes needed for the given string.
+        // Include an extra byte for the null terminator.
+        Span<byte> buffer = GetSpan(Encoding.UTF8.GetMaxByteCount(myString.Length) + 1, out bool mustCommit);
```
This `+ 1` for the null terminator can in theory overflow too.

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs:
```diff
@@ -138,12 +138,14 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
         HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks;
-        var pinnedBuffers = PinDataBuffers(endOfRequest, data, out dataChunks);
+
+        UnmanagedBufferAllocator allocator = new();
```
This will leak if there is an exception in BuildDataChunks

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/ResponseStreamAsyncResult.cs:
```diff
@@ -18,6 +18,7 @@ internal sealed unsafe partial class ResponseStreamAsyncResult : IAsyncResult, I
     private uint _bytesSent;
     private readonly CancellationToken _cancellationToken;
     private readonly CancellationTokenRegistration _cancellationRegistration;
+    private UnmanagedBufferAllocator _unmanagedAllocator;
```
Is this required for correctness to fix a pre-existing bug?

I am not able to review this for thread-safety. Can two different threads end up using the `_unmanagedAllocator` at the same time?

Also, given that this is async, it has higher chance of allocating the native memory on one thread and freeing it on a different thread. There is typically some penalty for this.

AaronRobinsonMSFT replied:
The threading here is non-trivial. I was going to take the simpler path, just create one at the callsite, but decided to encapsulate the logic inside the async result itself. The performance considerations are interesting here. As far as correctness (i.e., is the async result only accessed from a thread at a time), that was an assumption on my part. I can remove the member here instead.

AaronRobinsonMSFT committed:
Review feedback

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -277,31 +276,29 @@ private void EnsureResponseStream()
     // TODO: Consider using the HTTP_SEND_RESPONSE_FLAG_BUFFER_DATA flag for most/all responses rather than just Opaque.
-    internal unsafe uint SendHeaders(HttpApiTypes.HTTP_DATA_CHUNK[]? dataChunks,
+    internal unsafe uint SendHeaders(ref UnmanagedBufferAllocator allocator,
+        Span<HttpApiTypes.HTTP_DATA_CHUNK> dataChunks,
```
Did you also meant to allocate `HTTP_DATA_CHUNK`s in `BuildDataChunks` using `UnmanagedBufferAllocator`? 

(This change from array to span does not appear to be needed.)

AaronRobinsonMSFT replied:
Oh right. This was another option was exploring. It didn't work out since I would have needed to update several call sites and the changes weren't really all that beneficial.

jkotas replied:
It did not look too bad to me. I will leave it to @davidfowl to comment on how important it is to avoid managed allocations on this path. `HTTP_DATA_CHUNK[]` and `List<GCHandle>` can be switched to `UnmanagedBufferAllocator` and `Span<HTTP_DATA_CHUNK>` and `Span<GCHandle>`.

davidfowl replied:
We have similar code in IIS (https://github.com/dotnet/aspnetcore/blob/e35f2ea7994a3b18fc00c5f65c315739e41147be/src/Servers/IIS/IIS/src/Core/IO/AsyncWriteOperationBase.cs#L38-L53). I like the idea of switching this to unmanaged allocations too 👍🏾

AaronRobinsonMSFT committed:
Revert argument Span<> back to [].

AaronRobinsonMSFT committed:
Missed forgiveness operator.

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs:
```diff
@@ -188,7 +192,7 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
-    private List<GCHandle> PinDataBuffers(bool endOfRequest, ArraySegment<byte> data, out HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks)
+    private List<GCHandle> BuildDataChunks(ref UnmanagedBufferAllocator allocator, bool endOfRequest, ArraySegment<byte> data, out HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks)
     {
         var pins = new List<GCHandle>();
```
We should not need growable `List<GCHandle>' given the other changes. We know exactly how many GCHandles we are going to needed. We can allocate `GCHandle[]` of the exact size, similar how `HTTP_DATA_CHUNK[]` is allocated.

davidfowl commented:
BTW is there an easy way to track alloc/free using counters? Would it even be useful? Or would you switch to native profilers at that point?

jkotas commented:
You would switch to native profilers to profile unmanaged memory allocators.

AaronRobinsonMSFT commented:
> BTW is there an easy way to track alloc/free using counters? Would it even be useful? Or would you switch to native profilers at that point?

The VS profiler has an unmanaged memory allocation tracker. /cc @karpinsn

davidfowl commented on src/Servers/HttpSys/src/HttpSysListener.cs:
```diff
@@ -310,103 +309,79 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+        httpResponse.Response_V1.ReasonLength = checked((ushort)bytesLength);
+
+        int pContentLengthLength = 1;
+        byte* pContentLength = stackalloc byte[pContentLengthLength + 1];
```
stackalloc is more efficient with constants right?
```suggestion
        byte* pContentLength = stackalloc byte[2];
```

AaronRobinsonMSFT replied:
I've marked `pContentLengthLength` as `const` and with the `+ 1` this should be obvious to the JIT now.

jkotas replied:
I expect that the constant-prop optimization will be done in Roslyn when this is marked const. Constant-prop is one of the few optimizations done by Roslyn.

Rename `pContentLengthLength` to `contentLengthLength`? The variable is not a pointer, so it should not have `p` prefix.

davidfowl commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -527,128 +516,92 @@ private static bool Matches(string knownValue, StringValues input)
-                        gcHandle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
-                        pinnedHeaders.Add(gcHandle);
-                        pKnownHeaders[lookup].pRawValue = (byte*)gcHandle.AddrOfPinnedObject();
+                        HttpApiTypes.HTTP_RESPONSE_INFO* responseAlloc = allocator.Alloc< HttpApiTypes.HTTP_RESPONSE_INFO>(numKnownMultiHeaders);
```
nit: 

```suggestion
                        HttpApiTypes.HTTP_RESPONSE_INFO* responseAlloc = allocator.Alloc<HttpApiTypes.HTTP_RESPONSE_INFO>(numKnownMultiHeaders);
```

davidfowl commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -660,76 +613,65 @@ internal unsafe void SerializeTrailers(HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks
+        finally
         {
-            throw new HttpSysException((int)errorCode);
+            allocator.Dispose();
```
Any reason for an explicit dispose call here?

AaronRobinsonMSFT replied:
The `using` pattern isn't possible when taking a `ref` to the variable in the pattern. Note there is one place where the `using` was possible in this review.

davidfowl replied:
Add a comment?

davidfowl commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -660,76 +613,65 @@ internal unsafe void SerializeTrailers(HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks
-            true);
-
-        if (errorCode != ErrorCodes.ERROR_SUCCESS)
+        UnmanagedBufferAllocator allocator = new();
```
```suggestion
        using var allocator = new UnmanagedBufferAllocator();
```

AaronRobinsonMSFT replied:
Can't do that due to taking a `ref` to the `allocator`.

davidfowl commented on src/Shared/Buffers.MemoryPool/UnmanagedBufferAllocator.cs:
```diff
@@ -0,0 +1,157 @@
+    /// <typeparam name="T">The type requested</typeparam>
+    /// <param name="count">The count in <typeparamref name="T"/> units</param>
+    /// <returns>A pointer to the reserved memory.</returns>
+    public T* Alloc<T>(int count) where T : unmanaged
```
Is there any reason to return T* instead of `Span<T>`?

AaronRobinsonMSFT replied:
`Span<T>` is cumbersome in this sense because one isn't supposed to be able to access the pointer. I can return a `Span<T>`, but then we need to use the `Unsafe.AsPointer(ref MemoryMarshal.GetReference(...))` pattern in all places. This avoids those gymnastics. I can go back to that if desired.

AaronRobinsonMSFT replied:
I've updated this API to be `AllocAsPointer()` and created a new one, `AllocAsSpan()`. Now callers can decide the most appropriate function to call. Converting more allocations to use this allocator type uncovered places where always being a `Span<T>` is more appropriate.

davidfowl commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -527,128 +516,92 @@ private static bool Matches(string knownValue, StringValues input)
         }
 
-        try
+        fixed (HttpApiTypes.HTTP_KNOWN_HEADER* pKnownHeaders = &_nativeResponse.Response_V1.Headers.KnownHeaders)
```
@jkotas Would it be possible to replace this with ref locals?

jkotas replied:
Possible - yes. Beneficial - depends on your style preference, the short-term pinning is not a performance concern.

davidfowl replied:
I just experimented with it, so it's purely a style difference. Unsafe pointers vs managed pointers.

AaronRobinsonMSFT commented:
> Fixes/helps #40215?

Oh wow. Yes! So glad there is an issue filed for this. I'll mark this PR as fixing that.

AaronRobinsonMSFT commented on src/Servers/HttpSys/src/HttpSysListener.cs:
```diff
@@ -310,103 +309,79 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+        pContentLength[1] = 0; // null terminator
+
+        (&httpResponse.Response_V1.Headers.KnownHeaders)[(int)HttpSysResponseHeader.ContentLength].pRawValue = pContentLength;
+        (&httpResponse.Response_V1.Headers.KnownHeaders)[(int)HttpSysResponseHeader.ContentLength].RawValueLength = checked((ushort)pContentLengthLength);
```
Nope. This should be "1". The Headers length fields don't include the null terminator.

davidfowl commented:
I think this is a good first pass to measure performance. Some potential changes once we get through this:
- Managed pointers
- Stackalloc for known headers when under a certain size
- Stackalloc for data chunks when under a certain size

AaronRobinsonMSFT committed:
Update stackalloc usage

AaronRobinsonMSFT committed:
Convert more allocations to unmanged

AaronRobinsonMSFT committed:
Provide an Alloc for pointer or Span<T>

AaronRobinsonMSFT committed:
Fix spelling.

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs:
```diff
@@ -188,30 +192,41 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
+        var pins = allocator.AllocAsSpan<GCHandle>(maxGCHandleCount);
+
+        // Manually initialize the allocated GCHandles
+        for (int i = 0; i < maxGCHandleCount; ++i)
```
```suggestion
        for (int i = 0; i < pins.Length; ++i)
```
This will allow bounds-check elimination when accessing the Span to kick in.

jkotas replied:
Alternatively, you can do `pins.Clear()`.

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs:
```diff
@@ -188,30 +192,41 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
+        for (int i = 0; i < maxGCHandleCount; ++i)
+        {
+            pins[i] = new GCHandle();
+        }
```
It would make the method return `Span<HttpApiTypes.HTTP_DATA_CHUNK>`, pass `Span<GCHandle>` as out argument instead and assign the out argument here before we allocated any GCHandles.

It fixes the GCHandle memory leak in case there is an exception in the rest of this method.

AaronRobinsonMSFT commented:
Benchmark for current changes.

```
| application                             | baseline                 | Current                  |         |
| --------------------------------------- | ------------------------ | ------------------------ | ------- |
| CPU Usage (%)                           |                       64 |                       67 |  +4.69% |
| Cores usage (%)                         |                    1,788 |                    1,880 |  +5.15% |
| Working Set (MB)                        |                      484 |                      486 |  +0.41% |
| Private Memory (MB)                     |                      514 |                      530 |  +3.11% |
| Build Time (ms)                         |                    6,666 |                    3,886 | -41.70% |
| Start Time (ms)                         |                      328 |                      344 |  +4.88% |
| Published Size (KB)                     |                  118,659 |                  118,659 |   0.00% |
| .NET Core SDK Version                   | 8.0.100-alpha.1.22514.14 | 8.0.100-alpha.1.22514.14 |         |
| Max CPU Usage (%)                       |                       67 |                       65 |  -2.51% |
| Max Working Set (MB)                    |                      501 |                      508 |  +1.40% |
| Max GC Heap Size (MB)                   |                      255 |                      257 |  +0.73% |
| Size of committed memory by the GC (MB) |                      452 |                      456 |  +0.77% |
| Max Number of Gen 0 GCs / sec           |                     8.00 |                     7.00 | -12.50% |
| Max Number of Gen 1 GCs / sec           |                     2.00 |                     2.00 |   0.00% |
| Max Number of Gen 2 GCs / sec           |                     1.00 |                     1.00 |   0.00% |
| Max Time in GC (%)                      |                     1.00 |                     1.00 |   0.00% |
| Max Gen 0 Size (B)                      |               36,651,360 |               41,094,992 | +12.12% |
| Max Gen 1 Size (B)                      |                6,313,496 |                8,741,000 | +38.45% |
| Max Gen 2 Size (B)                      |                1,511,712 |                1,463,744 |  -3.17% |
| Max LOH Size (B)                        |                  113,184 |                  113,184 |   0.00% |
| Max POH Size (B)                        |                1,773,536 |                  825,936 | -53.43% |
| Max Allocation Rate (B/sec)             |            1,687,780,712 |            1,760,190,512 |  +4.29% |
| Max GC Heap Fragmentation               |                       90 |                       91 |  +1.19% |
| # of Assemblies Loaded                  |                       88 |                       88 |   0.00% |
| Max Exceptions (#/s)                    |                        0 |                        0 |         |
| Max Lock Contention (#/s)               |                    1,078 |                    1,400 | +29.87% |
| Max ThreadPool Threads Count            |                       48 |                       48 |   0.00% |
| Max ThreadPool Queue Length             |                      178 |                      134 | -24.72% |
| Max ThreadPool Items (#/s)              |                1,319,545 |                1,130,717 | -14.31% |
| Max Active Timers                       |                        0 |                        0 |         |
| IL Jitted (B)                           |                  117,381 |                  156,942 | +33.70% |
| Methods Jitted                          |                    1,315 |                    2,083 | +58.40% |


| load                   | baseline  | Current   |         |
| ---------------------- | --------- | --------- | ------- |
| CPU Usage (%)          |        40 |        41 |  +2.50% |
| Cores usage (%)        |     1,108 |     1,149 |  +3.70% |
| Working Set (MB)       |        49 |        48 |  -2.04% |
| Private Memory (MB)    |       376 |       376 |   0.00% |
| Start Time (ms)        |         0 |         0 |         |
| First Request (ms)     |       215 |       243 | +13.02% |
| Requests/sec           |   401,248 |   423,907 |  +5.65% |
| Requests               | 6,058,808 | 6,400,799 |  +5.64% |
| Mean latency (ms)      |      0.69 |      0.62 | -10.17% |
| Max latency (ms)       |     24.05 |     22.32 |  -7.19% |
| Bad responses          |         0 |         0 |         |
| Socket errors          |         0 |         0 |         |
| Read throughput (MB/s) |     63.52 |     67.11 |  +5.65% |
| Latency 50th (ms)      |      0.58 |      0.56 |  -3.76% |
| Latency 75th (ms)      |      0.78 |      0.74 |  -4.77% |
| Latency 90th (ms)      |      0.99 |      0.96 |  -3.03% |
| Latency 99th (ms)      |      2.20 |      1.72 | -21.82% |
```

AaronRobinsonMSFT committed:
Improve exception safety.

AaronRobinsonMSFT committed:
Remove unnecessary allocation

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs:
```diff
@@ -188,30 +192,45 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
+            pins = allocator.AllocAsSpan<GCHandle>(1);
+
+            // Manually initialize the allocated GCHandle
+            pins.Clear();
```
```suggestion
            pins[0] = default;
```

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs:
```diff
@@ -188,30 +192,45 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
 
-        var chunkCount = hasData ? 1 : 0;
+        const int maxGCHandleCount = 4;
+        pins = allocator.AllocAsSpan<GCHandle>(maxGCHandleCount);
```
You can move it further down and only allocate `chunkCount` work of GCHandles.

AaronRobinsonMSFT committed:
Remove the notion of max allocation and instead
always allocate precisely what is needed.

Fix an unsafe exception code path.

JamesNK commented on src/Servers/HttpSys/src/HttpSysListener.cs:
```diff
@@ -310,103 +309,79 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+        httpResponse.Response_V1.ReasonLength = checked((ushort)bytesLength);
+
+        const int contentLengthLength = 1;
+        byte* pContentLength = stackalloc byte[contentLengthLength + 1];
```
We've placed `[SkipLocalsInit]` on perf sensitive methods with stackalloc. All the values are set so could add it here. Although the stackalloc is so small, it might not be worth the trouble.

jkotas replied:
Alternatively, we can have pre-allocated blob in unmanaged memory - similar to https://github.com/dotnet/aspnetcore/pull/44561#discussion_r997706203.

AaronRobinsonMSFT replied:
I'll use the allocator for this.

jkotas commented on src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs:
```diff
@@ -229,52 +242,59 @@ private List<GCHandle> PinDataBuffers(bool endOfRequest, ArraySegment<byte> data
         else if (endOfRequest)
         {
             _requestContext.Response.MakeTrailersReadOnly();
         }
```
```suggestion
        }
        Debug.Assert(currentChunk == dataChunks.Length);
```

AaronRobinsonMSFT commented on src/Servers/HttpSys/src/RequestProcessing/Response.cs:
```diff
@@ -527,128 +516,92 @@ private static bool Matches(string knownValue, StringValues input)
     }
 
-    internal unsafe void SerializeTrailers(HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks, int currentChunk, List<GCHandle> pins)
+    internal unsafe void SerializeTrailers(ref UnmanagedBufferAllocator allocator, Span<HttpApiTypes.HTTP_DATA_CHUNK> dataChunks, int currentChunk)
```
I avoided this logic because of the assert within `SerializeTrailers`. My reading of this was the "trailers" must be at the end so the assert of that case should occur within that function as opposed to it's callers. Since there is only one caller I guess this is okay, but I'm not convinced this is worth the change as it spreads a seemingly important invariant to all potential call sites.

jkotas replied:
I have been reviewing the code to make sure that the exact right number of chunks is initialized. 

It made me to suggest the assert at the end https://github.com/dotnet/aspnetcore/pull/44561#discussion_r997732383 and also this change. As written, it is not clear whether `SerializeTrailers` fills in one dataChunk or multiple dataChunks.

AaronRobinsonMSFT commented on src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs:
```diff
@@ -188,30 +192,39 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
+            pins[0] = default;
+
+            dataChunks = allocator.AllocAsSpan<HttpApiTypes.HTTP_DATA_CHUNK>(chunkCount);
+            SetDataChunk(dataChunks, ref currentChunk, new ArraySegment<byte>(Helpers.ChunkTerminator), out pins[pinsIndex++]);
```
That is going to make things complicated. The first being I will need some sort of static location to perform that operation, perhaps a `static` constructor will work. I then will need to capture the length, which I could do with a `ReadOnlySpan<T>`, but these types are in a `static class`.  This means I could use `GC.AllocateArray()`, but I still need the static location. Do you have a preference?

I can do all of this but these optimizations are getting very narrow and this PR continues to grow. I think we need to agree on a reasonable stopping point here. In my mind this is starting to approach some very cumbersome code for minimal benefit.

AaronRobinsonMSFT replied:
I've applied this change.

AaronRobinsonMSFT committed:
Remove more GCHandles

AaronRobinsonMSFT committed:
Review feedback.
Remove explicit type static constructor.

AaronRobinsonMSFT commented on src/Servers/HttpSys/src/Helpers.cs:
```diff
@@ -2,14 +2,35 @@
 {
-    internal static readonly byte[] ChunkTerminator = new byte[] { (byte)'0', (byte)'\r', (byte)'\n', (byte)'\r', (byte)'\n' };
-    internal static readonly byte[] CRLF = new byte[] { (byte)'\r', (byte)'\n' };
+    // Both of these arrays are pinned.
+    private static readonly byte[] ChunkTerminator = CreatePinnedArray("0\r\n\r\n"u8);
+    private static readonly byte[] CRLF = CreatePinnedArray("\r\n"u8);
+
+    private static byte[] CreatePinnedArray(ReadOnlySpan<byte> data)
+    {
+        var array = GC.AllocateArray<byte>(data.Length, pinned: true);
+        data.CopyTo(array);
+        return array;
+    }
+
+    internal static unsafe void GetCRLF(out byte* ptr, out int length)
+    {
+        ptr = (byte*)Marshal.UnsafeAddrOfPinnedArrayElement(CRLF, 0);
+        length = CRLF.Length;
+    }
+
+    internal static unsafe void GetChunkTerminator(out byte* ptr, out int length)
+    {
+        ptr = (byte*)Marshal.UnsafeAddrOfPinnedArrayElement(ChunkTerminator, 0);
+        length = ChunkTerminator.Length;
+    }
```
> I would stay away from GC.AllocateArray(..., pinned: true). It is very hard to reason about its perf characteristics.

That makes me sad.

@davidfowl looks like we made the right choice ;)

jkotas replied:
We are not going to change UTF8 string literal implementation to return moveable strings. It would be unnatural change and code out there depends on the UTF8 string literals to be pinned. I would be ok with simplifying it to 

```
internal static class Helpers
{
    // We are assuming that the UTF8 string literals are implicitly pinned
    // See https://github.com/dotnet/csharplang/issues/5295
    public static ReadOnlySpan<byte> ChunkTerminator => "0\r\n\r\n"u8;
    public static ReadOnlySpan<byte> CRLF => "\r\n"u8;
```

AaronRobinsonMSFT replied:
The trouble with that is getting the pointer out of the underlying `ReadOnlySpan<T>`. I think for now I am going to leave this as is.

jkotas replied:
`Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedSpan))`. You need it in just one place here: https://github.com/dotnet/aspnetcore/pull/44561/files#diff-8279c4653ca902e3c8f3a3b68fef3fb12744e2b96cacc6f2215b86c712f750f7R298. (+change the method to take span and make the caller responsible for pinning the span).

AaronRobinsonMSFT replied:
> Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedSpan)). 

Yep, I was using that previously and have been trying to hide that implementation detail. Using the `fixed` keyword and then having it go out of scope is confusing in the best of times and in this case isn't clear this is permitted in this narrow instance. I also need it in https://github.com/dotnet/aspnetcore/pull/44561/files#diff-3e99422acb5996cdbb88be5ebbfa1d94e2c419f872da11f47f9fc5e90e835b43.

AaronRobinsonMSFT committed:
Review feedback

AaronRobinsonMSFT commented:
/cc @davidfowl 

# Benchmarks against .NET 8

```
| application                             | baseline                 | Current                  |         |
| --------------------------------------- | ------------------------ | ------------------------ | ------- |
| CPU Usage (%)                           |                       66 |                       67 |  +1.52% |
| Cores usage (%)                         |                    1,852 |                    1,883 |  +1.67% |
| Working Set (MB)                        |                      479 |                      473 |  -1.25% |
| Private Memory (MB)                     |                      505 |                      499 |  -1.19% |
| Build Time (ms)                         |                    6,445 |                    4,194 | -34.93% |
| Start Time (ms)                         |                      334 |                      355 |  +6.29% |
| Published Size (KB)                     |                  115,430 |                  115,430 |   0.00% |
| .NET Core SDK Version                   | 8.0.100-alpha.1.22514.14 | 8.0.100-alpha.1.22514.14 |         |
| Max CPU Usage (%)                       |                       68 |                       68 |  +0.68% |
| Max Working Set (MB)                    |                      502 |                      494 |  -1.57% |
| Max GC Heap Size (MB)                   |                      312 |                      332 |  +6.19% |
| Size of committed memory by the GC (MB) |                      451 |                      450 |  -0.25% |
| Max Number of Gen 0 GCs / sec           |                     6.00 |                     6.00 |   0.00% |
| Max Number of Gen 1 GCs / sec           |                     1.00 |                     1.00 |   0.00% |
| Max Number of Gen 2 GCs / sec           |                     0.00 |                     0.00 |         |
| Max Time in GC (%)                      |                     1.00 |                     1.00 |   0.00% |
| Max Gen 0 Size (B)                      |               46,443,032 |               25,880,352 | -44.28% |
| Max Gen 1 Size (B)                      |                6,664,304 |                6,506,072 |  -2.37% |
| Max Gen 2 Size (B)                      |                1,357,992 |                1,348,240 |  -0.72% |
| Max LOH Size (B)                        |                  113,184 |                  113,184 |   0.00% |
| Max POH Size (B)                        |                  839,320 |                  822,840 |  -1.96% |
| Max Allocation Rate (B/sec)             |            1,858,940,728 |            1,778,665,328 |  -4.32% |
| Max GC Heap Fragmentation               |                       92 |                       91 |  -1.51% |
| # of Assemblies Loaded                  |                       88 |                       88 |   0.00% |
| Max Exceptions (#/s)                    |                        0 |                        0 |         |
| Max Lock Contention (#/s)               |                    1,390 |                    1,306 |  -6.04% |
| Max ThreadPool Threads Count            |                       48 |                       48 |   0.00% |
| Max ThreadPool Queue Length             |                      138 |                      133 |  -3.62% |
| Max ThreadPool Items (#/s)              |                1,190,411 |                1,173,617 |  -1.41% |
| Max Active Timers                       |                        0 |                        0 |         |
| IL Jitted (B)                           |                  116,702 |                  154,697 | +32.56% |
| Methods Jitted                          |                    1,316 |                    2,064 | +56.84% |


| load                   | baseline  | Current   |         |
| ---------------------- | --------- | --------- | ------- |
| CPU Usage (%)          |        42 |        42 |   0.00% |
| Cores usage (%)        |     1,186 |     1,180 |  -0.51% |
| Working Set (MB)       |        49 |        49 |   0.00% |
| Private Memory (MB)    |       376 |       376 |   0.00% |
| Start Time (ms)        |         0 |         0 |         |
| First Request (ms)     |       240 |       239 |  -0.42% |
| Requests/sec           |   431,565 |   433,502 |  +0.45% |
| Requests               | 6,516,506 | 6,545,631 |  +0.45% |
| Mean latency (ms)      |      0.75 |      0.66 | -11.59% |
| Max latency (ms)       |     23.46 |     32.78 | +39.73% |
| Bad responses          |         0 |         0 |         |
| Socket errors          |         0 |         0 |         |
| Read throughput (MB/s) |     68.32 |     68.63 |  +0.45% |
| Latency 50th (ms)      |      0.55 |      0.55 |  +0.37% |
| Latency 75th (ms)      |      0.73 |      0.73 |  -0.82% |
| Latency 90th (ms)      |      0.97 |      0.95 |  -2.06% |
| Latency 99th (ms)      |      8.44 |      2.81 | -66.71% |
```

AaronRobinsonMSFT commented:
/cc @davidfowl 

# Benchmarks against .NET 8

```
| application                             | baseline                 | Current                  |         |
| --------------------------------------- | ------------------------ | ------------------------ | ------- |
| CPU Usage (%)                           |                       65 |                       72 | +10.77% |
| Cores usage (%)                         |                    1,816 |                    2,014 | +10.90% |
| Working Set (MB)                        |                      469 |                      472 |  +0.64% |
| Private Memory (MB)                     |                      495 |                      498 |  +0.61% |
| Build Time (ms)                         |                    3,746 |                    3,667 |  -2.11% |
| Start Time (ms)                         |                      382 |                      360 |  -5.76% |
| Published Size (KB)                     |                  115,434 |                  115,434 |   0.00% |
| .NET Core SDK Version                   | 8.0.100-alpha.1.22514.14 | 8.0.100-alpha.1.22514.14 |         |
| Max CPU Usage (%)                       |                       64 |                       75 | +16.34% |
| Max Working Set (MB)                    |                      489 |                      494 |  +1.01% |
| Max GC Heap Size (MB)                   |                      333 |                      334 |  +0.29% |
| Size of committed memory by the GC (MB) |                      441 |                      448 |  +1.60% |
| Max Number of Gen 0 GCs / sec           |                     6.00 |                     6.00 |   0.00% |
| Max Number of Gen 1 GCs / sec           |                     1.00 |                     1.00 |   0.00% |
| Max Number of Gen 2 GCs / sec           |                     0.00 |                     0.00 |         |
| Max Time in GC (%)                      |                     1.00 |                     0.00 |         |
| Max Gen 0 Size (B)                      |               46,476,784 |               33,895,144 | -27.07% |
| Max Gen 1 Size (B)                      |                6,111,888 |                6,507,688 |  +6.48% |
| Max Gen 2 Size (B)                      |                1,354,112 |                1,341,880 |  -0.90% |
| Max LOH Size (B)                        |                  113,184 |                  113,184 |   0.00% |
| Max POH Size (B)                        |                  835,200 |                  826,960 |  -0.99% |
| Max Allocation Rate (B/sec)             |            1,837,387,536 |            1,743,068,656 |  -5.13% |
| Max GC Heap Fragmentation               |                       94 |                       91 |  -3.98% |
| # of Assemblies Loaded                  |                       88 |                       88 |   0.00% |
| Max Exceptions (#/s)                    |                        0 |                        0 |         |
| Max Lock Contention (#/s)               |                    1,357 |                    1,285 |  -5.31% |
| Max ThreadPool Threads Count            |                       48 |                       48 |   0.00% |
| Max ThreadPool Queue Length             |                      137 |                      134 |  -2.19% |
| Max ThreadPool Items (#/s)              |                1,154,303 |                1,122,875 |  -2.72% |
| Max Active Timers                       |                        0 |                        0 |         |
| IL Jitted (B)                           |                  157,134 |                  154,667 |  -1.57% |
| Methods Jitted                          |                    2,082 |                    2,064 |  -0.86% |


| load                   | baseline  | Current   |         |
| ---------------------- | --------- | --------- | ------- |
| CPU Usage (%)          |        42 |        42 |   0.00% |
| Cores usage (%)        |     1,185 |     1,165 |  -1.69% |
| Working Set (MB)       |        49 |        49 |   0.00% |
| Private Memory (MB)    |       376 |       376 |   0.00% |
| Start Time (ms)        |         0 |         0 |         |
| First Request (ms)     |       235 |       237 |  +0.85% |
| Requests/sec           |   429,490 |   423,980 |  -1.28% |
| Requests               | 6,484,821 | 6,401,831 |  -1.28% |
| Mean latency (ms)      |      0.70 |      0.67 |  -4.55% |
| Max latency (ms)       |     22.15 |     37.66 | +70.02% |
| Bad responses          |         0 |         0 |         |
| Socket errors          |         0 |         0 |         |
| Read throughput (MB/s) |     67.99 |     67.12 |  -1.28% |
| Latency 50th (ms)      |      0.55 |      0.56 |  +1.64% |
| Latency 75th (ms)      |      0.73 |      0.75 |  +1.63% |
| Latency 90th (ms)      |      0.97 |      0.99 |  +2.06% |
| Latency 99th (ms)      |      5.50 |      2.83 | -48.55% |
```

AaronRobinsonMSFT committed:
Review feedback

AaronRobinsonMSFT committed:
Remove unused using.

jkotas commented:
LGTM!

AaronRobinsonMSFT commented:
@davidfowl @Tratcher @JamesNK Please take another look. The CI is all green.

davidfowl commented:
So, it's a bit slower or a wash and the GC numbers look about the same. 🤔

AaronRobinsonMSFT commented:
Sharing offline comments I made based on collection of a nettrace for `GCHandle` usage.

I'm convinced on this change. The logging itself is a win. The number of GCHandle events is staggering. I did a quick experiment locally. Instead of using `NativeMemory.Alloc` and `NativeMemory.Free`, I replaced them with `ArrayPool<byte>.Shared.Rent` and a pinned `GCHandle`. The trace resulted in precisely the same experience before my larger changes. This is a win and the performance characteristics that we've improved aren't reflected in the crank base statistics. They are however reflected in nettrace.

The nettrace impact is prior to this change the nettrace was 1GB, but after this change it is approximately 40MB. The win here is the substantial impact of `GCHandle` events.

davidfowl commented:
@sebastienros I want to assign this to you. Once we have benchmarks running on .NET 8 we should merge this.

sebastienros commented:
@davidfowl we are tracking net8.0 now (check the charts) so I assume that's what you wanted before we merge the PR.

davidfowl commented:
Yep

davidfowl commented on src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:
```diff
@@ -19,12 +16,7 @@ internal static unsafe string GetString(byte* pBytes, int byteCount, bool useLat
         else
         {
-            return new ReadOnlySpan<byte>(pBytes, byteCount).GetAsciiOrUTF8StringNonNullCharacters(Encoding);
+            return new ReadOnlySpan<byte>(pBytes, byteCount).GetAsciiOrUTF8StringNonNullCharacters(Encoding.UTF8);
```
Are these the same settings as Encoding.UTF8? I imagine there's a reason we had a custom instantiation.

AaronRobinsonMSFT replied:
The original `Encoding` was also `UTF8`, but explicitly set `encoderShouldEmitUTF8Identifier: false` which is the only semantic difference from `Encoding.UTF8`. I was deferring to @jkotas's suggestion, see https://github.com/dotnet/aspnetcore/pull/44561#discussion_r996475101. Please let me know if you'd prefer to change it back.

BrennanConroy commented:
> The nettrace impact is prior to this change the nettrace was 1GB, but after this change it is approximately 40MB. The win here is reduction of the impact of `GCHandle` events.

Could you share what nettrace args are needed to show this? I'm thinking ahead to next year when we write a ASP.NET Core 8 perf improvements blog post and it's useful to have info on the PR so we don't need to figure out how to repro things a year later.

AaronRobinsonMSFT commented:
@BrennanConroy This is easy to see by adding the following arguments to the `crank` command: `--application.dotnetTrace true --application.dotnetTraceProviders gc-verbose`. Running scenarios where a lot of `GCHandle`s instance are created/destroyed will saturate the pipeline and result in dropped events or at best, very large nettrace files.
